Tree Closing and Opening

Rhythm1710 opened the pull request:
Fixes: https://github.com/rust-lang/bors/issues/207

Kobzol commented on src/bors/handlers/queue.rs:
```diff
@@ -0,0 +1,58 @@
+
+/// Process the queue of pull requests for each repository
+#[allow(dead_code)]
+pub async fn process_queue(
```
This file is still present in the PR, it should be removed.

Rhythm1710 replied:
I will remove this.

Kobzol commented on src/bors/handlers/review.rs:
```diff
@@ -486,12 +549,17 @@ PR will need to be re-approved."#,
 approve = ["+approved"]
 "#,
         );
+        repo.lock()
```
Why was this added here?

Rhythm1710 replied:
I think this was added when I was resolving conflicts. I will remove this.

Kobzol commented on src/bors/handlers/review.rs:
```diff
@@ -610,6 +678,61 @@ approve = ["+approved"]
+                    .db()
+                    .get_repository_treeclosed(&default_repo_name())
+                    .await?;
+                if repo_models.is_empty() {
```
The test shouldn't modify the DB in this way, just read from it.

Kobzol commented on src/bors/handlers/review.rs:
```diff
@@ -156,6 +156,68 @@ pub(super) async fn handle_push_to_pull_request(
+        return Ok(());
+    }
+
+    db.update_repository_treeclosed(repo_state.repository(), 0)
```
Let's create two functions, `close_tree` and `open_tree`, on the DB. Let the DB layer decide how it will store the representation of closed trees.

Rhythm1710 replied:
Should we notify about tree open and tree closed?

Rhythm1710 replied:
Should we notify about tree open and tree closed in this layer only?

Kobzol replied:
I'm not sure what do you mean by that. The notification is implemented in the command handler, which is where it should be, IMO.

Kobzol commented on src/database/mod.rs:
```diff
@@ -83,8 +83,44 @@ impl sqlx::Type<sqlx::Postgres> for PullRequestNumber {
+    }
+}
+
+impl From<Option<TreeState>> for TreeState {
```
This is a pretty weird impl, it's very unobvious if someone uses `.into()` to go from `Option<TreeState>` to `TreeState` and it magically selects this default through this impl. This should be set explicitly somewhere else.

Kobzol commented on src/database/mod.rs:
```diff
@@ -101,7 +137,7 @@ pub enum BuildStatus {
 
 /// Represents a single (merged) commit.
-#[derive(Debug, sqlx::Type)]
+#[derive(Debug, Clone, sqlx::Type)]
```
What are the clones needed for?

Rhythm1710 replied:
I was getting error in the try_build of Pull_Request_Model for not using clone in Build_Model.

Rhythm1710 replied:
![Uploading image.png…]()

Kobzol replied:
Sure, but then why `PullRequestModel` needs to be clone now? :D

Kobzol reviewed (commented):
Thanks for the changes. Left a few more comments.

I wonder if we should pre-fill the DB with known repositories when bors starts, because it does know the list of repositories at the beginning. Then we wouldn't have to assume everywhere in code that the repository might be missing. Although there are some edge cases around a repository being added to bors while it is running (although we will ~never hit that in practice). Anyway, that can be done in a separate PR.

Rhythm1710 commented:
> Thanks for the changes. Left a few more comments.
> 
> I wonder if we should pre-fill the DB with known repositories when bors starts, because it does know the list of repositories at the beginning. Then we wouldn't have to assume everywhere in code that the repository might be missing. Although there are some edge cases around a repository being added to bors while it is running (although we will ~never hit that in practice). Anyway, that can be done in a separate PR.

> I think that is a better approach. If you want I can add this in this PR only .

Kobzol commented:
I rebased the PR and refactored some parts a bit. I left one TODO for you to resolve, to fill in the source of the tree closed message. It should be a URL pointing to the PR comment that closed the tree.

Rhythm1710 commented:
> I rebased the PR and refactored some parts a bit. I left one TODO for you to resolve, to fill in the source of the tree closed message. It should be a URL pointing to the PR comment that closed the tree.

Implemented this using

Function:- `source: format!(
                "https://github.com/{}/pull/{}",
                repo_state.repository(),
                pr.number
            ),`

Test:-`source: format!(
                            "https://github.com/{}/pull/{}",
                            default_repo_name(),
                            default_pr_number()
                        ),`

Kobzol commented:
We want to have a link to the specific comment that closed the tree, not just the PR, that could have a lot of comments (we want to see the reason why the tree was closed).
The `issue_comment` webhook contains a property `comment.html_url`, which contains a HTML URL link to the comment. It should be parsed from the webhook and stored in the parsed pull request.

We can then use that as the tree closed source.

Kobzol commented:
Thank you!
